### PR TITLE
Add models

### PR DESCRIPTION
**Pull request by znicholls**
Adds two-layer model and impulse response model. Also includes a notebook which goes through the details of why the two-layer and impulse response models are identical. 

- [x] Tests added
- [x] Documentation added
- [x] Example added (in the documentation, to an existing notebook, or in a new notebook)
- [x] Description in ``CHANGELOG.rst`` added (single line such as: ``(`#XX <https://github.com/openscm/openscm-twolayermodel/pull/XX>`_) Added feature which does something``)

For reference:

- two-layer code already here https://github.com/znicholls/mullet/blob/prototype/sandbox/speed-test-multiple-parameters.py
- impulse response code here https://github.com/OMS-NetZero/FAIR/blob/c189beadf68a14bd18bd7b8b9604ee9b780c8a08/fair/forward.py#L155

### CODE REVIEW

**Comment by znicholls** [comment]
@rgieseke this is ready to have a first look at (there's a bug somewhere, something for later/tomorrow). Probably start with https://github.com/openscm/openscm-twolayermodel/blob/c84c8a4e2b4c6deab077a9f463486e1cf8589c23/notebooks/impulse-response-equivalence.ipynb then go from there during a reivew?

File: src/openscm_twolayermodel/constants.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+"""
+Physical constatns used in calculations
```
Comment (rgieseke) [inline src/openscm_twolayermodel/constants.py]:
```suggestion
Physical constants used in calculations
```

File: src/openscm_twolayermodel/base.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    """
+    Base class for model implementations
+    """
+
+    _save_paras = tuple()  # parameters to save when doing a run
+
+    _name = None  # model name
+
+    @staticmethod
+    def _check_is_pint_quantity(quantity, name, model_units):
```
Comment (rgieseke) [inline src/openscm_twolayermodel/base.py]: I'd probably call this
`_is_pint_quantity` or `_check_if_pint_quantity`

File: src/openscm_twolayermodel/impulse_response_model.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            )
+
+            if str(efficacy_term.units) != "watt / meter ** 2":
+                raise AssertionError("units should have come out as W/m^2")
+
+        out = erf - lambda_0.magnitude * (t1 + t2) - efficacy_term.magnitude
+
+        return out
+
+    def _get_run_output_tss(self, ts_base):
```
Comment (rgieseke) [inline src/openscm_twolayermodel/impulse_response_model.py:246]: "tss" is just plural for "ts" as timeseries?

**Review by rgieseke — COMMENTED** [review]
Great job! 
Given that there quite a few variants in the literature it seems, the object-oriented design should be useful!

Minor consistency thing, there are a few variables like `lambda_0` or `b_pt_1` with underscores before the number, most others without, maybe unify?

File: src/openscm_twolayermodel/impulse_response_model.py (hunk 3 above)
Comment (znicholls) [reply src/openscm_twolayermodel/impulse_response_model.py:246]: yep, nastily named but internal so maybe ok?

Comment (rgieseke) [reply src/openscm_twolayermodel/impulse_response_model.py:246]: sure! maybe add a comment in the base class?

File: src/openscm_twolayermodel/base.py (hunk 2 above)
Comment (znicholls) [reply src/openscm_twolayermodel/base.py]: Have updated, resolve if happy

**Comment by znicholls** [comment]
> Minor consistency thing, there are a few variables like `lambda_0` or `b_pt_1` with underscores before the number, most others without, maybe unify?

Done

File: src/openscm_twolayermodel/impulse_response_model.py (hunk 3 above)
Comment (znicholls) [reply src/openscm_twolayermodel/impulse_response_model.py:246]: done

**Comment by rgieseke** [comment]
What is the 't' prefix for variables, just for testing?

**Comment by rgieseke** [comment]
All green! Should we merge this now?

**Comment by znicholls** [comment]
Yep t for test.

I will wrangle pylint for 20 minutes, then we can merge.

**Pull request merged**